File: ChatSupport/src/main/java/com/vantagecircle/chatapp/holder/UserMViewHolder.java
package com.vantagecircle.chatapp.holder;

import android.support.v7.widget.RecyclerView;
import android.util.Log;
import android.view.View;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.vantagecircle.chatapp.R;
import com.vantagecircle.chatapp.Support;
import com.vantagecircle.chatapp.core.DataModel;
import com.vantagecircle.chatapp.core.GetDataHandler;
import com.vantagecircle.chatapp.core.interfacep.ChildInterface;
import com.vantagecircle.chatapp.core.interfacep.ResultInterface;
import com.vantagecircle.chatapp.utils.ConfigUtils;
import com.vantagecircle.chatapp.utils.Constant;
import com.vantagecircle.chatapp.interfacePref.ClickUser;
import com.vantagecircle.chatapp.model.ChatM;
import com.vantagecircle.chatapp.model.UserM;

/**
 * Created by bapidas on 27/07/17.
 */

public class UserMViewHolder extends RecyclerView.ViewHolder implements View.OnClickListener {
    private TextView user_name, email_id, last_message;
    private LinearLayout sub_holder, lastImage;
    private ClickUser clickUser;

    UserMViewHolder(View itemView) {
        super(itemView);
        user_name = (TextView) itemView.findViewById(R.id.user_name);
        email_id = (TextView) itemView.findViewById(R.id.email_id);
        sub_holder = (LinearLayout) itemView.findViewById(R.id.sub_holder);
        lastImage = (LinearLayout) itemView.findViewById(R.id.lastImage);
        last_message = (TextView) itemView.findViewById(R.id.last_message);
        sub_holder.setOnClickListener(this);
    }

    public void setViewHolder(UserM userM, ClickUser clickUser) {
        if (!Support.id.equals(userM.getUserId())) {
            sub_holder.setVisibility(View.VISIBLE);
            itemView.setVisibility(View.VISIBLE);
            this.clickUser = clickUser;
            user_name.setText(userM.getFullName());
            email_id.setText(userM.getUsername());

            getLastMessage(userM);
        } else {
            sub_holder.setVisibility(View.GONE);
            itemView.setVisibility(View.GONE);
        }
    }

    private void getLastMessage(final UserM userM) {
        ConfigUtils.checkRooms(userM, new ResultInterface() {
            @Override
            public void onSuccess(String t) {
                if(t.equals(Constant.NO_ROOM)){
                    last_message.setVisibility(View.GONE);
                    lastImage.setVisibility(View.GONE);
                } else {
                    bindLastMessage(t);
                }
            }

            @Override
            public void onFail(String e) {
                Log.d("onFail", "Error " + e);
            }
        });
    }

    private void bindLastMessage(String room){
        GetDataHandler getDataHandler1 = new GetDataHandler();
        getDataHandler1.setQueryReference(Support.getChatReference()
                .child(room)
                .orderByKey()
                .limitToLast(1));
        getDataHandler1.setChildValueListener(new ChildInterface() {
            @Override
            public void onChildNew(DataModel dataModel) {
                ChatM chatM = dataModel.getDataSnapshot().getValue(ChatM.class);
                assert chatM != null;
                if(chatM.getChatType().equals(Constant.TEXT_TYPE)){
                    lastImage.setVisibility(View.GONE);
                    last_message.setVisibility(View.VISIBLE);
                    last_message.setText(chatM.getMessageText());
                } else {
                    lastImage.setVisibility(View.VISIBLE);
                    last_message.setVisibility(View.GONE);
                }
            }

            @Override
            public void onChildModified(DataModel dataModel) {
                ChatM chatM = dataModel.getDataSnapshot().getValue(ChatM.class);
                assert chatM != null;
                if(chatM.getChatType().equals(Constant.TEXT_TYPE)){
                    lastImage.setVisibility(View.GONE);
                    last_message.setVisibility(View.VISIBLE);
                    last_message.setText(chatM.getMessageText());
                } else {
                    lastImage.setVisibility(View.VISIBLE);
                    last_message.setVisibility(View.GONE);
                }
            }

            @Override
            public void onChildDelete(DataModel dataModel) {

            }

            @Override
            public void onChildRelocate(DataModel dataModel) {

            }

            @Override
            public void onChildCancelled(DataModel dataModel) {
                Log.d("onChildCancelled", "Error " + dataModel.getDatabaseError().getMessage());
            }
        });
    }

    @Override
    public void onClick(View v) {
        if (v.getId() == R.id.sub_holder) {
            if (clickUser != null) {
                clickUser.onUserClick(getAdapterPosition());
            }
        }
    }
}

File: app/src/main/java/com/bapidas/chattingapp/ui/adapter/holder/UserMViewHolder.kt
package com.bapidas.chattingapp.ui.adapter.holder

import android.util.Log
import android.view.View
import android.widget.LinearLayout
import android.widget.TextView
import androidx.recyclerview.widget.RecyclerView
import com.bapidas.chattingapp.ChatApplication
import com.bapidas.chattingapp.R
import com.bapidas.chattingapp.data.core.GetDataHandler
import com.bapidas.chattingapp.data.core.callbacks.ChildInterface
import com.bapidas.chattingapp.data.core.callbacks.ResultInterface
import com.bapidas.chattingapp.data.core.model.DataModel
import com.bapidas.chattingapp.data.model.ChatM
import com.bapidas.chattingapp.data.model.UserM
import com.bapidas.chattingapp.ui.adapter.callbacks.ClickUser
import com.bapidas.chattingapp.utils.ConfigUtils.checkRooms
import com.bapidas.chattingapp.utils.Constants

/**
 * Created by bapidas on 27/07/17.
 */
class UserMViewHolder constructor(itemView: View) : RecyclerView.ViewHolder(itemView), View.OnClickListener {
    private val userName: TextView = itemView.findViewById<View>(R.id.user_name) as TextView
    private val emailId: TextView = itemView.findViewById<View>(R.id.email_id) as TextView
    private val lastMessage: TextView = itemView.findViewById<View>(R.id.last_message) as TextView
    private val subHolder: LinearLayout = itemView.findViewById<View>(R.id.sub_holder) as LinearLayout
    private val lastImage: LinearLayout = itemView.findViewById<View>(R.id.lastImage) as LinearLayout

    private var clickUser: ClickUser? = null

    override fun onClick(v: View) {
        if (v.id == R.id.sub_holder) {
            if (clickUser != null) {
                clickUser?.onUserClick(adapterPosition)
            }
        }
    }

    init {
        subHolder.setOnClickListener(this)
    }

    fun setViewHolder(userM: UserM, clickUser: ClickUser?) {
        if (ChatApplication.applicationContext().id != userM.userId) {
            subHolder.visibility = View.VISIBLE
            itemView.visibility = View.VISIBLE
            this.clickUser = clickUser
            userName.text = userM.fullName
            emailId.text = userM.username
            getLastMessage(userM)
        } else {
            subHolder.visibility = View.GONE
            itemView.visibility = View.GONE
        }
    }

    private fun getLastMessage(userM: UserM) {
        checkRooms(userM, object : ResultInterface {
            override fun onSuccess(t: String) {
                if (t == Constants.NO_ROOM) {
                    lastMessage.visibility = View.GONE
                    lastImage.visibility = View.GONE
                } else {
                    bindLastMessage(t)
                }
            }

            override fun onFail(e: String) {
                Log.d("onFail", "Error $e")
            }
        })
    }

    private fun bindLastMessage(room: String) {
        val getDataHandler1 = GetDataHandler()
        val ref = ChatApplication.applicationContext().chatReference
                .child(room).orderByKey().limitToLast(1)
        getDataHandler1.setChildValueListener(ref, object : ChildInterface {
            override fun onChildNew(dataModel: DataModel) {
                val chatM = dataModel.dataSnapshot?.getValue(ChatM::class.java)
                if (chatM != null) {
                    if (chatM.chatType == Constants.TEXT_CONTENT) {
                        lastImage.visibility = View.GONE
                        lastMessage.visibility = View.VISIBLE
                        lastMessage.text = chatM.messageText
                    } else {
                        lastImage.visibility = View.VISIBLE
                        lastMessage.visibility = View.GONE
                    }
                }
            }

            override fun onChildModified(dataModel: DataModel) {
                val chatM = dataModel.dataSnapshot?.getValue(ChatM::class.java)
                if (chatM != null) {
                    if (chatM.chatType == Constants.TEXT_CONTENT) {
                        lastImage.visibility = View.GONE
                        lastMessage.visibility = View.VISIBLE
                        lastMessage.text = chatM.messageText
                    } else {
                        lastImage.visibility = View.VISIBLE
                        lastMessage.visibility = View.GONE
                    }
                }
            }

            override fun onChildDelete(dataModel: DataModel) {

            }
            override fun onChildRelocate(dataModel: DataModel) {

            }
            override fun onChildCancelled(dataModel: DataModel) {
                Log.d("onChildCancelled", "Error " + dataModel.databaseError?.message)
            }
        })
    }
}